locator/src/control_plane.rs

```diff
@@ -0,0 +1,161 @@
+#![allow(dead_code)]
+
+use crate::types::{Cell, CellId, RouteData};
+use reqwest::{StatusCode, Url};
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::time::{Duration, sleep};
+
+#[derive(Deserialize)]
+struct ControlPlaneRecord {
+    id: String,
+    slug: String,
+    cell: CellId,
+}
+
+#[derive(Deserialize)]
+struct ControlPlaneMetadata {
+    cursor: String,
+    has_more: bool,
+    cell_to_locality: HashMap<String, String>,
+}
+
+#[derive(Deserialize)]
+struct ControlPlaneData {
+    data: Vec<ControlPlaneRecord>,
+    metadata: ControlPlaneMetadata,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ControlPlaneError {
+    #[error("could not load config: {0}")]
+    ReqwestError(#[from] reqwest::Error),
+    #[error("invalid URL: {0}")]
+    InvalidUrl(String),
+    #[error("control plane unavailable")]
+    ControlPlaneRetriesExceeded,
+}
+
+// A cursor is passed for incremental loading. No cursor means the full snapshot will be loaded.
+pub async fn load_mappings(
+    base_url: &str,
+    cursor: Option<&str>,
+) -> Result<RouteData, ControlPlaneError> {
+    const RETRIABLE_STATUS_CODES: &[StatusCode] = &[
+        StatusCode::TOO_MANY_REQUESTS,     // 429
+        StatusCode::INTERNAL_SERVER_ERROR, // 500
+        StatusCode::BAD_GATEWAY,           // 502
+        StatusCode::SERVICE_UNAVAILABLE,   // 503
+        StatusCode::GATEWAY_TIMEOUT,       // 504
+    ];
+
+    let mut cells: HashMap<String, Arc<Cell>> = HashMap::new();
+    let mut org_to_cell = HashMap::new();
+    let mut next_cursor: Option<String> = cursor.map(String::from);
+    let mut page_fetches = 0;
+    // TODO: consider reusing client
+    let client = reqwest::Client::new();
+
+    // 3 retries per page fetch
+    let mut retries = 0;
+
+    loop {
+        let mut url =
+            Url::parse(base_url).map_err(|e| ControlPlaneError::InvalidUrl(e.to_string()))?;
+
+        if let Some(ref c) = next_cursor {
+            url.query_pairs_mut().append_pair("cursor", c);
+        }
+
+        let response = client.get(url).send().await?;
+
+        if !response.status().is_success() {
+            if RETRIABLE_STATUS_CODES.contains(&response.status()) && retries < 3 {
+                retries += 1;
+                // Backoff between retries
+                let retry_millis = 2_u64.pow(retries);
```

Comment:
Should we increase this from 2ms to something more viable like 10 or 100ms? Since there are only 3 retries, the maximum backoff we would have is 8ms which seems quite small.

Maybe put these magic number as constants at the very top so its easy to reason about them.